ZIP 316: support larger Unified Addresses and Unified Viewing Keys (#551)
zcash/zips: zip-0316.rst

  unambiguously with a specific Receiver Type, identified by an integer Typecode.
Receiver Encoding
  An encoding of a Receiver as a byte sequence.

> Viewing Key

Reviewer: Done in 6756428f760a512a7b02969c9fb594f0a43243c0

>   The necessary information to view information about a payments to an Address,
>   or (in the case of a Full Viewing Key) from an Address.

Reviewer: ```suggestion
  The necessary information to view information about payments to an Address,
  or (in the case of a Full Viewing Key) from an Address.
```

>   The necessary information to view information about a payments to an Address,
>   or (in the case of a Full Viewing Key) from an Address.

Reviewer 2: ```suggestion
  The necessary information to view information about payments to an Address,
  or (in the case of a Full Viewing Key) from an Address. An Incoming Viewing
  Key can be derived from a Full Viewing Key, and an Address can be derived
  from an Incoming Viewing Key.
```
This change was lost in a force-push.

  shielded support without re-acquiring counterparty UAs. If they are
  re-acquired, the user flow and usability will be minimally disrupted.

Experimental Usage
------------------

Unified Addresses and Unified Viewing Keys must be able to include
Receivers and Viewing Keys of experimental types, possibly alongside
non-experimental ones. These experimental Receivers or Viewing Keys
must be used only by wallets whose users have explicitly opted into
the corresponding experiment.

Viewing Keys
------------

> 

Reviewer: > Therefore, a UFVK or UIVK should be able to include a Transparent Address.

str4d: We can't do diversification for transparent keys, ends up duplicating the same t-key across diversified shielded sapling/orchard keys, linking them. We need a 'stream of unlinkable addresses' , (extended pubkey)

daira: have to be separate derivation trees for transparent/sapling/orchard because they all use different curves

You can't have a ViewingKey that includes P2SH.

Author: I will resolve this.

:math:`(\mathtt{typecode}, \mathtt{length}, \mathtt{addr})` encodings
of the consituent Receivers:

* :math:`\mathtt{typecode} : \mathtt{compactSize}` — the Typecode from the
  above Priority List;

> * :math:`\mathtt{length} : \mathtt{compactSize}` — the length in bytes of

Reviewer: TODO(daira): this needs to be updated to be below the lower limit of zcashd's, that limits the length of compact size encoding to the number of elements (length encoding vs number encoding)

Unified Full or Incoming Viewing Keys are encoded and decoded
analogously to Unified Addresses. A Consumer MUST use the decoding
procedure from the previous section. For Viewing Keys, a Consumer
will normally take the union of information provided by all contained
Receivers, and therefore the Priority List defined in the previous
section is not used.

For Receiver Types that are currently defined for UFVKs and UIVKs in
this specification, the same Typecodes are used as for a corresponding

> Unified Address. Additional Receiver Types for UFVKs and UIVKs MAY be

Reviewer: TODO(@daira): make plural agreement consistent

defined in future, and these will not necessarily use the same Typecode
as the corresponding Unified Address.

The following UFVK or UIVK Encodings are used in place of the
:math:`\mathtt{addr}` field:

> * An Orchard UFVK or UIVK Encoding, with Type Code :math:`\mathtt{0x03}`,
>   is the raw encoding of the Full Viewing Key or Incoming Viewing Key
>   respectively.
> 

Reviewer: ```suggestion
* An Orchard UFVK or UIVK Encoding, with Type Code :math:`\mathtt{0x03}`,
  is the raw encoding of the Orchard Full Viewing Key or Orchard Incoming Viewing Key
  respectively.

```

* A Sapling UFVK Encoding, with Type Code :math:`\mathtt{0x02}`, is
  the encoding of a Sapling Extended Full Viewing Key defined in

>   [#zip-0032-sapling-extfvk]_. This SHOULD be a full viewing key at

Reviewer: This 'SHOULD' is not Sapling-specific, re: HD encodings

> * A Sapling UFVK Encoding, with Type Code :math:`\mathtt{0x02}`, is
>   the encoding of a Sapling Extended Full Viewing Key defined in
>   [#zip-0032-sapling-extfvk]_. This SHOULD be a full viewing key at
>   the account level.

Reviewer: ```suggestion
* A Sapling UFVK Encoding, with Type Code :math:`\mathtt{0x02}`, is
  the encoding of a Sapling Extended Full Viewing Key defined in
  [#zip-0032-sapling-extfvk]_. This SHOULD be a Full Viewing Key at
  the account level.
```

* A Sapling UIVK Encoding, also with Type Code :math:`\mathtt{0x02}`,

>   is an encoding of :math:`(\mathsf{dk}, \mathsf{ivk}` given by:

Reviewer: ```suggestion
  is an encoding of :math:`(\mathsf{dk}, \mathsf{ivk})` given by:
```

>   :math:`I2LEOSP_{88}(\mathsf{dk}) || I2LEOSP_{256}(\mathsf{ivk})`

Reviewer: ```suggestion
  :math:`\mathsf{I2LEOSP}_{88}(\mathsf{dk}) || \mathsf{I2LEOSP}_{256}(\mathsf{ivk})`
```

> * There is no defined way to represent a viewing key for a Transparent

Reviewer: ```suggestion
* There is no defined way to represent a Viewing Key for a Transparent
```

Author: Explain why.

  that cannot be interpreted as specified above.

Adding new types
----------------

It is intended that new Receiver Types and Viewing Key Types SHOULD be
introduced either by a modification to this ZIP or by a new ZIP, in accordance
with the ZIP Process [#zip-0000]_.

Experimental types MAY be added using the reserved Typecodes 0xFFFA to 0xFFFF
inclusive. This provides for six simultaneous experiments, which can be
referred to as experiments A to F. This should be sufficient because
experiments are expected to be reasonably short-term, and should otherwise
be either standardized in a ZIP (and allocated a Typecode outside this
reserved range) or discontinued.

> 

Author: That was @str4d's idea. Yes I like it too!

> 

Author: On reflection I think the wording is clear as-is. There's nothing to stop experimental type users from coordinating to avoid collisions, so it wouldn't be correct to say they should *expect* them.

  that cannot be interpreted as specified above.

Adding new types
----------------

It is intended that new Receiver Types and Viewing Key Types SHOULD be
introduced either by a modification to this ZIP or by a new ZIP, in accordance
with the ZIP Process [#zip-0000]_.

> Experimental types MAY be added using the reserved Typecodes 0xFFFA to 0xFFFF

Author: It's unclear whether this requires a ZIP, given the wording above.

Suggestion: move most of this section up to where the typecodes are defined, and say here "for experimentation prior to proposing a ZIP".

  that cannot be interpreted as specified above.

Adding new types
----------------

It is intended that new Receiver Types and Viewing Key Types SHOULD
be introduced either by a modification to this ZIP or by a new ZIP,
in accordance with the ZIP Process [#zip-0000]_.

Experimental types MAY be added using the reserved Typecodes
:math:`\mathtt{0xFFFA}` to :math:`\mathtt{0xFFFF}` inclusive. This
provides for six simultaneous experiments, which can be referred to
as experiments A to F. This should be sufficient because experiments
are expected to be reasonably short-term, and should otherwise be
either standardized in a ZIP (and allocated a Typecode outside this
reserved range) or discontinued.

Deriving a UIVK from a UFVK
---------------------------

Each UIVK Encoding can straightforwardly be derived from the
corresponding UFVK Encoding, without changing the Type Code. In the
case of a Transparent P2PKH UFVK Encoding, the UIVK Encoding is the
same.

> Deriving a Unified Address from a UIVK

Reviewer: TODO(@daira ) : check if FF1 is a S-PRP, want to avoid a possibly linkability attack if keys diversified in serial

Author: As far as I know this is not a significant concern. The best attacks on FF1 (e.g. up to 7 rounds out of 10 in https://link.springer.com/chapter/10.1007/978-3-319-63715-0_23) are only practical for much smaller domain sizes than 2<sup>88</sup>, so in practice there is a large safety margin that should be sufficient to protect against chosen ciphertext attacks.

in accordance with the ZIP Process [#zip-0000]_.

For experimentation prior to proposing a ZIP, experimental types MAY
be added using the reserved Typecodes :math:`\mathtt{0xFFFA}` to
:math:`\mathtt{0xFFFF}` inclusive. This provides for six simultaneous
experiments, which can be referred to as experiments A to F. This
should be sufficient because experiments are expected to be reasonably
short-term, and should otherwise be either standardized in a ZIP (and
allocated a Typecode outside this reserved range) or discontinued.

Deriving a UIVK from a UFVK
---------------------------

Each UIVK Encoding can straightforwardly be derived from the
corresponding UFVK Encoding, without changing the Typecode. In the
case of a Transparent P2PKH UFVK Encoding, the UIVK Encoding is the
same.

Deriving a Unified Address from a UIVK
--------------------------------------

To derive a Unified Address from a UIVK we need to choose a diversifier
index, which MUST be valid for all of the Viewing Key Types in the
UIVK. That is,

* A Sapling diversifier index MUST generate a valid diversifier as
  defined in ZIP 32 section “Sapling diversifier derivation”
  [#zip-0032-sapling-diversifier-derivation]_.

> * A Transparent diversifier index MUST be in the range :math:`0` to

Reviewer 2: nb: should this refer specifically to a P2PKH address, or is "transparent" sufficient in conjunction with the exclusion around P2SH addresses for UIVKs?

Author: I think this is sufficient.

* let :math:`y = a \oplus H_0(x)`
* let :math:`d = x \oplus G_1(y)`
* let :math:`c = y \oplus H_1(d)`
* return :math:`c \,||\, d.`

The inverse function :math:`\mathsf{F4Jumble}^{-1}` is obtained in the usual
way for a Feistel construction, by observing that :math:`r = p \oplus q` implies :math:`p = r \oplus q.`

The first argument to BLAKE2b below is the personalization.

We instantiate :math:`H_i(u)` by
:math:`\mathsf{BLAKE2b‐}(8\ell_L)(\texttt{“UA_F4Jumble_H”} \,||\,`
:math:`[i, 0, 0], u).`

We instantiate :math:`G_i(u)` as the first :math:`\ell_R` bytes of the
concatenation of

> :math:`[\mathsf{BLAKE2b‐}512(\texttt{“UA_F4Jumble_G”} \,||\, [i] \,||\,`

Author: It's a sequence of the single value i, which fits in a byte (so there is no endianness issue). This is the same notation as in the protocol spec; I'll add a note to that effect.

> :math:`[\mathsf{BLAKE2b‐}512(\texttt{“UA_F4Jumble_G”} \,||\, [i] \,||\,`

Author: Done in 938841d91f57fd4b8cc25e06563fd7d8c735ceb1.

before encoding the result with Bech32m.

The Consumer rejects any Bech32m-decoded byte sequence that is less than

> 48 bytes or greater than 4194368 bytes; otherwise it applies

Reviewer 2: Where does this constant arise from? Could this constant be named and associated with the function that derives the bound?

Reviewer: TODO (@daira )

Address) means that this is also the maximum length as of NU5 activation.

For longer UAs (when other Typecodes are added), the cost increases to 6
BLAKE2b compressions for :math:`128 < \ell_M \leq 192,` and 10 BLAKE2b
compressions for :math:`192 < \ell_M \leq 256,` for example. The maximum
cost for which the algorithm is defined would be 196608 BLAKE2b compressions
at :math:`\ell_M = 4194368` bytes.

A naïve implementation of the :math:`F4Jumble^{-1}` function would require
roughly :math:`\ell_M` bytes plus the size of a BLAKE2b hash state. However,
it is possible to reduce this by streaming the :math:`d` part of the jumbled
encoding three times from a less memory-constrained device. It is essential
that the streamed value of :math:`d` is the same on each pass, which can be
verified using a MAC (with key held only by the Consumer) or
collision-resistant hash function. After the first pass of :math:`d`, the
implementation is able to compute :math:`y`; after the second pass it is
able to compute :math:`a`; and the third allows it to compute and

> incrementally parse :math:`b.` The maximum memory usage during this process
> would be 128 bytes (for the cached :math:`c` or :math:`y`), plus two BLAKE2b
> hash states.

Reviewer: ```suggestion
incrementally parse :math:`b.` The maximum memory usage during this process
would be 128 bytes (for :math:`a`, and either :math:`c` or :math:`y`), plus two BLAKE2b
hash states.
```

> would be 128 bytes (for the cached :math:`c` or :math:`y`), plus two BLAKE2b
> hash states.

Reviewer: ```suggestion
would be 128 bytes (for the cached :math:`c` or :math:`y`), plus 
two BLAKE2b hash states.
```

Since this streaming implementation of :math:`F4Jumble^{-1}` is quite
complicated, we do not require all Consumers to support it. If a Consumer
implementation cannot support UAs / UVKs up to the maximum length, it MUST

> nevertheless support UAs / UVKs with :math:`\ell_M` of at least 256 bytes.

Reviewer: Having two maximum lengths supported indicates two conformance levels
for UAs/UVKs: minimum for non-experimental types and the maximum including 
experimental types.

Author: I will clarify this.

Address) means that this is also the maximum length as of NU5 activation.

For longer UAs (when other Typecodes are added), the cost increases to 6
BLAKE2b compressions for :math:`128 < \ell_M \leq 192,` and 10 BLAKE2b
compressions for :math:`192 < \ell_M \leq 256,` for example. The maximum
cost for which the algorithm is defined would be 196608 BLAKE2b compressions
at :math:`\ell_M = 4194368` bytes.

A naïve implementation of the :math:`F4Jumble^{-1}` function would require
roughly :math:`\ell_M` bytes plus the size of a BLAKE2b hash state. However,
it is possible to reduce this by streaming the :math:`d` part of the jumbled
encoding three times from a less memory-constrained device. It is essential

> that the streamed value of :math:`d` is the same on each pass, which can be
> verified using a MAC (with key held only by the Consumer) or
> collision-resistant hash function. After the first pass of :math:`d`, the

Reviewer: ```suggestion
that the streamed value of :math:`d` is the same on each pass, which can be
verified using a Message Authentication Code (with key held only by the Consumer) or
collision-resistant hash function. After the first pass of :math:`d`, the
```

> implementation is able to compute :math:`y`; after the second pass it is
> able to compute :math:`a`; and the third allows it to compute and
> incrementally parse :math:`b.` The maximum memory usage during this process
> would be 128 bytes (for the cached :math:`c` or :math:`y`), plus two BLAKE2b
> hash states.

Reviewer: ```suggestion
implementation is able to compute :math:`y`; after the second pass it is
able to compute :math:`a`; and the third allows it to compute and
incrementally parse :math:`b.` The maximum memory usage during this process
would be 128 bytes (for the cached :math:`c` or :math:`y`, and also for :math:`a`),
plus two BLAKE2b hash states.
```

Since this streaming implementation of :math:`F4Jumble^{-1}` is quite

> complicated, we do not require all Consumers to support it. If a Consumer

Reviewer: ```suggestion
complicated, we do not require all Consumers to support streaming. If a Consumer
```
